browser/ui/brave_rewards/tip_panel_coordinator.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_UI_BRAVE_REWARDS_TIP_PANEL_COORDINATOR_H_
+#define BRAVE_BROWSER_UI_BRAVE_REWARDS_TIP_PANEL_COORDINATOR_H_
+
+#include <string>
+
+#include "base/memory/weak_ptr.h"
+#include "base/observer_list.h"
+#include "base/scoped_observation.h"
+#include "brave/components/brave_rewards/common/mojom/ledger_types.mojom.h"
+#include "chrome/browser/ui/browser_user_data.h"
+#include "ui/gfx/geometry/size.h"
+
+namespace brave_rewards {
+
+class RewardsService;
+
+// Provides a browser-scoped communication channel for components that need to
+// display the tip panel and components responsibly for showing the tip panel.
+class TipPanelCoordinator : public BrowserUserData<TipPanelCoordinator> {
+ public:
+  TipPanelCoordinator(Browser* browser, RewardsService* rewards_service);
+
+  TipPanelCoordinator(const TipPanelCoordinator&) = delete;
+  TipPanelCoordinator& operator=(const TipPanelCoordinator&) = delete;
+
+  ~TipPanelCoordinator() override;
+
+  // Displays the tip panel for the specified publisher.
+  void ShowPanelForPublisher(const std::string& publisher_id);
+
+  class Observer : public base::CheckedObserver {
+   public:
+    // Called when an application component requests the tip panel.
+    virtual void OnTipPanelRequested(const std::string& publisher_id) {}
+  };
+
+  void AddObserver(Observer* observer);
+  void RemoveObserver(Observer* observer);
+  using Observation = base::ScopedObservation<TipPanelCoordinator, Observer>;
+
+  // Retrieves the publisher ID associated with the most recent tip panel
+  // request.
+  const std::string& publisher_id() const { return publisher_id_; }
+
+  // Gets or sets the size of the browser in which the tip panel will be
+  // displayed. These dimensions can be used by the front-end to adjust layout
+  // appropriately.
+  const gfx::Size& browser_size() const { return browser_size_; }
+  void set_browser_size(const gfx::Size& size) { browser_size_ = size; }
+
+ private:
+  friend class BrowserUserData<TipPanelCoordinator>;
+
+  void GetUserTypeCallback(const std::string& publisher_id,
```

Comment:
I think the convention is to name callbacks after their actions (`On` + action). Can we follow that and name these `OnGetUserType` and `OnIsPublisherRegistered`?

Reply:
I'm not entirely sure what our convention should be. Just adding an `On` prefix for these async getters results in things like `OnGetUserType`, which sounds like it's an event handler for when a user type is requested, rather than a callback for the async getter.

I could use a weak-self lambda though 😉

Reply:
I do see examples of the "Callback" naming pattern (e.g. [here](https://source.chromium.org/chromium/chromium/src/+/main:components/payments/content/payment_request.h;l=207;bpv=1;bpt=1?q=%22Callback(%22%20lang:cpp&ss=chromium%2Fchromium%2Fsrc:components%2F&start=11))), but I'm not sure how common the "On" pattern is in comparison.